kernel/environ.mli

```diff
@@ -319,7 +313,7 @@ val remove_hyps : Id.Set.t -> (Context.Named.Declaration.t -> Context.Named.Decl
 open Retroknowledge
 (** functions manipulating the retroknowledge 
     @author spiwack *)
-val retroknowledge : (retroknowledge->'a) -> env -> 'a
+val retroknowledge : env -> retroknowledge
```

Comment:
@aspiwack what was the reason for writing this function in this way?

Reply:
Oh! It was a very nice reason! An important reason! A reason that will really impress you! A re…

I don't actually remember. Maybe I thought it was a very clever type for some reason? Maybe I had a particular pattern in mind. Maybe I just didn't realise that I was overcomplicating things.

Reply:
@aspiwack You seem to like this kind of pattern, I remember that you introduced a similar one for global proof data access. Let me dare a bit of software psychanalysis: I suppose that you're thinking in a richer language where the `'a` return type is constrained to have some property (e.g. `hProp`) so that the accessor looks like a higher induction principle. Somewhere in between your brain and the keyboard this gets compiled on the fly to the unexpressive OCaml language...

Reply:
Sure. Let's say that. At this point, I'll take any excuse.

Reply:
>@aspiwack You seem to like this kind of pattern, I remember that you introduced a similar one for global proof data access. Let me dare a bit of software psychanalysis: I suppose that you're thinking in a richer language where the 'a return type is constrained to have some property (e.g. hProp) so that the accessor looks like a higher induction principle. Somewhere in between your brain and the keyboard this gets compiled on the fly to the unexpressive OCaml language...

Wouldn't it make more sense to constrain in such a way that the access is global, ie the retroknowledge doesn't leak beyond the function call? A bit like the old Sigma stuff, or the haskell local state monad.

Reply:
@SkySkimmer OCaml has side-effects! Boo, boo, that breaks almost of the parametricity reasoning...